### PR TITLE
EC2 update: Systemd, module blacklisting, and nofail in fstab for cloud-init

### PR DESCRIPTION
**Pull request by JamesBromberger**

### CODE REVIEW

**Comment by andsens** [comment]
Crap, I started commenting on the original commits instead of this PR (only \06da3098 so far, hope you can live with that?).

File: bootstrapvz/common/tasks/grub.py [hunk 1]
```diff
@@ -27,9 +27,9 @@ def run(cls, info):
 		grub_def = os.path.join(info.root, 'etc/default/grub')
 		sed_i(grub_def, '^#GRUB_TERMINAL=console', 'GRUB_TERMINAL=console')
 		sed_i(grub_def, '^GRUB_CMDLINE_LINUX_DEFAULT="quiet"',
-		                'GRUB_CMDLINE_LINUX_DEFAULT="console=hvc0"')
+		                'GRUB_CMDLINE_LINUX_DEFAULT="init=/bin/systemd console=hvc0 console=ttyS0"')
```
Comment (andsens) [inline bootstrapvz/common/tasks/grub.py]: We should probably start commenting what all this does/effect it has :-)

wouldn't the second `console=` overwrite the first one?

File: bootstrapvz/common/tasks/grub.py [hunk 2]
```diff
@@ -27,9 +27,9 @@ def run(cls, info):
...
 		sed_i(grub_def, '^GRUB_TIMEOUT=[0-9]+', 'GRUB_TIMEOUT=0\n'
-		                                        'GRUB_HIDDEN_TIMEOUT=true')
+		                                        'GRUB_HIDDEN_TIMEOUT_QUIET=true')
```
Comment (andsens) [inline bootstrapvz/common/tasks/grub.py]: See [here](https://github.com/andsens/bootstrap-vz/commit/77cbb64b9f6ed743edb0255ca50fdb1e7c496d4c).

File: bootstrapvz/common/tasks/grub.py [hunk 3]
```diff
@@ -78,7 +78,7 @@ def run(cls, info):
 class InstallGrub_2(Task):
 	description = 'Installing grub 2'
 	phase = phases.system_modification
-	predecessors = [filesystem.FStab]
+	predecessors = [ConfigureGrub]
```
Comment (andsens) [inline bootstrapvz/common/tasks/grub.py]: Wow, that's a serious oversight. `update-grub` and `grub-install` depend heavily on what's in the defaults. This should be done for 1.99 as well.

File: bootstrapvz/plugins/cloud_init/__init__.py [hunk 4]
```diff
@@ -26,6 +26,8 @@ def resolve_tasks(taskset, manifest):
 		taskset.add(tasks.SetUsername)
 	if 'disable_modules' in options:
 		taskset.add(tasks.DisableModules)
+        if 'enable_modules' in options:
```
Comment (andsens) [inline bootstrapvz/plugins/cloud_init/__init__.py]: Indentation seems to be off here. Check with tox.

File: bootstrapvz/providers/ec2/tasks/ami.py [hunk 5]
```diff
@@ -125,3 +126,8 @@ def run(cls, info):
 			registration_params['sriov_net_support'] = 'simple'
 
 		info._ec2['image'] = info._ec2['connection'].register_image(**registration_params)
+		log = logging.getLogger(__name__)
+		if info._ec2['image']:
+			log.info("AMI has ID %s" % info._ec2['image'])
```
Comment (andsens) [inline bootstrapvz/providers/ec2/tasks/ami.py]: Use `format()` or string concatenation.

File: bootstrapvz/providers/ec2/tasks/network.py [hunk 6]
```diff
@@ -2,6 +2,8 @@
 from bootstrapvz.common import phases
 from bootstrapvz.common.tasks import kernel
 import os.path
+import os
+import logging
```
Comment (andsens) [inline bootstrapvz/providers/ec2/tasks/network.py]: ?

File: manifests/official/ec2/ebs-jessie-amd64-hvm.yml [hunk 7]
```diff
@@ -14,20 +14,40 @@ image:
...
-  bootloader: extlinux
+  bootloader: grub
   charmap: UTF-8
   locale: en_US
   timezone: UTC
 volume:
   backing: ebs
   partitions:
-    type: none
+    type: msdos
```
Comment (andsens) [inline manifests/official/ec2/ebs-jessie-amd64-hvm.yml]: Why not gpt?

File: manifests/official/ec2/ebs-jessie-amd64-hvm.yml [hunk 8]
```diff
@@ -14,20 +14,40 @@ image:
...
     root:
       filesystem: ext4
       size: 8GiB
 packages:
   mirror: http://cloudfront.debian.net/debian
+  components:
+  - main
+  - contrib
+  - non-free
+  install:
```
Comment (andsens) [inline manifests/official/ec2/ebs-jessie-amd64-hvm.yml]: I assume we truly need all these packages to get cloud init working?

File: bootstrapvz/common/tasks/grub.py (hunk 1 above)
Comment (JamesBromberger) [inline bootstrapvz/common/tasks/grub.py]: Should send console to both, from what I have been told by EC2 team; but that could be wrong. Seems to be working for me now though.

File: bootstrapvz/plugins/cloud_init/__init__.py (hunk 4 above)
Comment (JamesBromberger) [inline bootstrapvz/plugins/cloud_init/__init__.py]: Yep, tabs/spaces. Will send an update.

File: manifests/official/ec2/ebs-jessie-amd64-hvm.yml (hunk 8 above)
Comment (JamesBromberger) [inline manifests/official/ec2/ebs-jessie-amd64-hvm.yml]: CLI, boto are to make the instance usable without requiring every boto to have to load these on launch; if an instance starts within an IAM Role for API keys and a security group that is highly restrictive (or indeed, a subnet within a VPC with no Internet access) then bootstrapping these basic libraries and utils becomes difficult. Eg, imagine if an instance is nly permitted (by an policy) that it is only allowed to make outbound HTTPS connections; then apt-transport-https is a hard dependency). We've shipped this for some time to enable the base image be subjected to stricter run-time environments.

**Comment by andsens** [comment]
Ping me when you have fixed the merge conflicts :-)

**Comment by justinsb** [comment]
I'm trying to build a Debian Jessie image for Kubernetes - see https://github.com/kubernetes/kubernetes/pull/21226 - we have to build our own image to enable the cgroup memory controller.

It appears that the official Debian 8.1 image was built using this PR (e.g. including awscli) but that this PR was never merged.  Is that right, or is the actual bootstrap-vz template somewhere else?

**Comment by myhro** [comment]
Hi Justin,

I didn't read your entire PR out there because it is quite big and there's an ongoing discussion. But yes, [you are right][1], James build the images using his own fork and then send his changes to us.

Can you try the new (8.3) images that [were released yesterday][2]? My guess is that they will [solve many of your problems][3].

Regards,
Tiago.

[1]: https://github.com/kubernetes/kubernetes/pull/21226#issuecomment-186489229
[2]: https://lists.debian.org/debian-cloud/2016/02/msg00070.html
[3]: https://lists.debian.org/debian-cloud/2016/02/msg00065.html

**Comment by justinsb** [comment]
Thanks for the confirmation @myhro .  I actually thought this (andsens/bootstrap-vz) was the official repo, and wasn't aware that the fork was important until recently.  The actual image manifest itself is pretty small (https://github.com/kubernetes/kubernetes/pull/21226/files), though we're having lots of discussion about it.

I'd like to build images that are as close to the official images as possible, so I will base them off the 8.3 images that were just released.  But we have to enable the cgroups memory controller.  While we're there, we choose to preinstall some packages, but that is just an optimization.  The other thing we get is the potential for a better understood, more immutable image if we re-release it often.

Now I'm not sure whether I should be asking questions on this PR or on the fork or on the mailing list.  Perhaps you can repoint me if needed.  I was able to install the `cloud-initramfs-growroot` package and had no problem with resizing the root (though only with partition type msdos).  @JamesBromberger  what's the downside of that approach vs backporting?

**Comment by JamesBromberger** [comment]
On 20/02/2016 9:30 PM, Justin Santa Barbara wrote:
>
> Thanks for the confirmation @myhro <https://github.com/myhro> . I
> actually thought this (andsens/bootstrap-vz) was the official repo,
> and wasn't aware that the fork was important until recently. The
> actual image manifest itself is pretty small
> (https://github.com/kubernetes/kubernetes/pull/21226/files), though
> we're having lots of discussion about it.
>
> I'd like to build images that are as close to the official images as
> possible, so I will base them off the 8.3 images that were just
> released. But we have to enable the cgroups memory controller. While
> we're there, we choose to preinstall some packages, but that is just
> an optimization. The other thing we get is the potential for a better
> understood, more immutable image if we re-release it often.
>
> Now I'm not sure whether I should be asking questions on this PR or on
> the fork or on the mailing list. Perhaps you can repoint me if needed.
> I was able to install the |cloud-initramfs-growroot| package and had
> no problem with resizing the root (though only with partition type
> msdos). @JamesBromberger <https://github.com/JamesBromberger> what's
> the downside of that approach vs backporting?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/andsens/bootstrap-vz/pull/256#issuecomment-186596708>.
>


The two repos (Anders' versus my repo) are pretty much the same - I feed
changes back to Anders via PRs; and we've tried to stay in sync over
time (I think we're 9 commits different right ahead/behind now). My
opinion is that Anders' version is the project.

The content of the manifest per platform - is subject to the
DDs/DMs/sub-team per Cloud platform; some platforms require various
pieces for core infrastructure pre-existing in the code base, or perhaps
customer kernels. Furthermore there are some base useful pieces that
would be recommended minimum installs as well. From a security point of
view, I like the idea of offering apt-transport-https in the base image
(as the Ec2 images do) as it permits you to design a base image that
boots and can *only* use HTTPS outbound form the start - an important
security requirement for some (ie, disable outbound port 80 completely)
- and without the chicken and egg of having to get apt-transport-https
into an instance that can' t make http requests to get the
apt-transport-https package (yes, you could do a curl to a well known
location for the package, and then install that, but here its already
done for you). I have generally asked on list for input from anyone
around - such as when we decided on the initial username that users SSH
to instances on....


-- 
/Mobile:/ +61 422 166 708, /Email:/ james_AT_rcpt.to

**Comment by myhro** [comment]
Justin,

> I actually thought this (andsens/bootstrap-vz) was the official repo, and wasn't aware that the fork was important until recently.

This is indeed the main repository for the project, as James has said, albeit we are not perfectly in sync with his fork all the time. 

> Now I'm not sure whether I should be asking questions on this PR or on the fork or on the mailing list. Perhaps you can repoint me if needed.

I would say that it depends. Here we can discuss about any issues regarding bootstrap-vz, how it relates to third-party tools it uses and the like. I would recommend the `debian-cloud` mailing list for issues regarding official published images (e.g. not about building your own) or Debian and its packages itself.

**Comment by andsens** [comment]
Hi James.
I have ported some of your changes. Could you rebase and send the PR again. It's pretty much only the tuning and cloud-init stuff that's missing.
e.g.: The grub options should be working properly now.

**Comment by andsens** [comment]
Nevermind, I did that as well now :-D

I think upstream is now up-to-date with your fork James, could you check?

**Pull request closed without merging**